### PR TITLE
Draft implementation of @properties for locks & exceptions

### LINKED ISSUE #32: Timeouts on get_ and set_ functions with decorators?
How do we set timeouts universally? Should we use decorators? Could use something along these lines:
```
import signal
import time
from contextlib import contextmanager

class TimeoutException(Exception): pass

@contextmanager
def time_limit(seconds):
    def signal_handler(signum, frame):
        raise TimeoutException("Timed out!")
    signal.signal(signal.SIGALRM, signal_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)

def timelimitdecorator(func):
    def wrapper(*args, **kwargs):
        try:
            with time_limit(10):
                func(*args, **kwargs)
        except TimeoutException as e:
            print("Timed out!")
    return wrapper
```

Then on a function
```
@timelimitdecorator
def foo(sleeptime):
    time.sleep(sleeptime)
    print("hello")
```

We get a timeout on foo(11), but not on foo(1).

### LINKED ISSUE #35: Locks handled as @properties and a careful programming proposition
**Note: this is something I think we should save effort on until we get a working prototype out, but want to write it down**

Reading through the code in a bit more detail while setting up alarms, and some of this locking is making me nervous. Please understand my intentions as genuine and goode and not trying to code shame. I raised this during our last call and just wanted to follow up

**Problem:**
locks, well, lock. and so if a lock is ever acquired without being released, and there are no timeouts, then the program hangs. The following code makes me nervous:

https://github.com/CohenLabPrinceton/Ventilator-Dev/blob/32d64d05f622e445633393847e63ef8afbfd11f9/vent/coordinator/coordinator.py#L142
```python
while not self.stopping.is_set():
    sensor_values = self.control_module.get_sensors()
    self.lock.acquire()
    self.COPY_sensor_values = sensor_values
    self.last_message_timestamp = sensor_values.timestamp
    self.lock.release()
    for name in [ValueName.PIP,
                 ValueName.PIP_TIME,
                 ValueName.PEEP,
                 ValueName.BREATHS_PER_MINUTE,
                 ValueName.INSPIRATION_TIME_SEC]:
        self.lock.acquire()
        in_control_settings = name not in self.COPY_control_settings
        self.lock.release()
        if in_control_settings:
            control_setting = self.control_module.get_control(name)
            self.lock.acquire()
            self.COPY_control_settings[name] = control_setting
            self.lock.release()

        self.lock.acquire()
        disagreed_tentative = name in self.COPY_tentative_control_settings and self.COPY_tentative_control_settings[name] != self.COPY_control_settings[
                    name]
        self.lock.release()
        if disagreed_tentative:
            self.lock.acquire()
            tentative_control_setting = self.COPY_tentative_control_settings[name]
            self.lock.release()
            self.control_module.set_control(tentative_control_setting)
            self.lock.acquire()
            self.COPY_control_settings[name] = self.control_module.get_control(name)
            self.lock.release()
    # sleep 10 ms
    time.sleep(0.01)
```

Given the number of times the lock is accessed, and the degree of nesting, this seems like a pretty difficult to maintain/error-prone method of thread management. 

**Proposal:**
From what I understand, the lock is just acquired and released whenever any of these set of attributes are accessed. 

let's use properties, this is what they were born to do.

Observe the following extraordinarily careful pattern that I believe we should follow as a general pattern for repetitively called methods

```python
import traceback
import functools

# some global stack param
MAX_STACK_DEPTH = 20


def locked(func):
    """
    Wrapper to use as decorator, handle lock logic for a
    @property

    Arguments:
        func (callable): function to wrap
    """

    # define a function to return that wraps the inner
    # function around some standardized error handling
    # logic

    # use .wraps to preserve inner method identity
    @functools.wraps(func)
    def lock_wrapper(self, *args, **kwargs):
        ret = None

        try:
            self.lock.acquire()
            ret = func(self, *args, **kwargs)
        except Exception as e:
            log_exception(
                e, 
                traceback.extract_tb(MAX_STACK_DEPTH)
            )
        finally:
            self.lock.release()

        return ret

    return lock_wrapper

class CoordinatorBase:
    def __init__(self, sim_mode=False):

        self.lock = threading.Lock()

        self.control_module = get_control_module(sim_mode)
        self._sensor_values = None

    @locked
    @property
    def sensor_values(self):
        return self._sensor_values

    @locked
    @sensor_values.setter
    def sensor_values(self, sensor_values):
        self._sensor_values = sensor_values
```

This would allow every lock acquisition and release to be replaced with standard attribute set/get methods. ie instead of
```
# set
self.lock.acquire()
self.COPY_sensor_values = ...
self.lock.release()

# get
def get_sensors(self) -> SensorValues:
    self.lock.acquire()
    sensor_values = self.COPY_sensor_values
    self.lock.release()
    # return res
    return sensor_values
```
(which is also retrieved by an additonal method `get_sensors` that acquires/releases the lock) 

you just use
```
# set
self.sensor_values = ...

# get
self.sensor_values
```

The benefits are of course not just aesthetic, but guarantee that the lock is always released due to the use of `finally` and that any exceptions are always logged.

### LINKED ISSUE #38: Error handling in vent.io
Need to make sure all possible exceptions & hangs are caught.

See: [This](https://github.com/CohenLabPrinceton/Ventilator-Dev/pull/31#discussion_r418366771) and 
[this.](https://github.com/CohenLabPrinceton/Ventilator-Dev/pull/31#discussion_r418367429)

### PR DESCRIPTION
**Pull request by CommReteris**
This is a rough draft - mostly untested and I didn't go through the project to apply @locked. exception logger doesn't actually log anything, just raises whatever it's passed.

Nevertheless, I figured I'd open up a feature branch & draft PR to collect thoughts & code. 

closes #32 
closes #35 
closes #38

### CODE REVIEW

File: vent/common/fashion.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+""" Decorators for dangerous functions
+
+"""
+from vent.common.values import MAX_STACK_DEPTH
+from vent.common.logging import log_exception
+import traceback
+import functools
+
+
+def locked(func):
```
Comment (sunnyszy) [inline vent/common/fashion.py:10]: Maybe I'm not very up to date. Did you use this anyway in the code?

File: vent/common/values.py [hunk 2]
```diff
@@ -318,3 +318,4 @@ def __getitem__(self, key):
 whether (``True``) or not (``False``) a value is allowed.
 """
 
+MAX_STACK_DEPTH = 20
```
Comment (sunnyszy) [inline vent/common/values.py:321]: Better to put this in logging.py, as this only related to logging.

File: vent/common/fashion.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    # use .wraps to preserve inner method identity
+    @functools.wraps(func)
+    def lock_wrapper(self, *args, **kwargs):
+        ret = None
+
+        try:
+            self.lock.acquire()
+            ret = func(self, *args, **kwargs)
+        except Exception as e:
+            log_exception(
```
Comment (sunnyszy) [inline vent/common/fashion.py:32]: What upper layer exception handling mechanism do you expect? E.g., restart the program, or just retry the reading?

Comment (sneakers-the-rat) [reply vent/common/fashion.py:32]: i think this is something that the coordinator would handle -- exceptions would trigger different actions depending on the state of the system. eg. if we have some max duration to wait for a new sensor reading, missing a reading below that duration shouldn't trigger anything but after that duration should throw an alarm.

Comment (sunnyszy) [reply vent/common/fashion.py:32]: Yeah, this is a good example~

Comment (CommReteris) [reply vent/common/fashion.py:32]: We definitely have to think through the different exceptions and what actions can/should be taken for each. Another one that comes to mind is when the Pigpio daemon can't be reached, the coordinator should kill and restart it.

Comment (sneakers-the-rat) [reply vent/common/fashion.py:32]: idk why we should overcomplicate it more than defining a few exception types and mapping them to a few predefined actions -- like make a `PigpioConnectionException` and throw it up to the coordinator who matches exception classes to actions.

File: vent/common/values.py (hunk 2 above)
Comment (sneakers-the-rat) [reply vent/common/values.py:321]: Ya I think there will need to be a separate user/system-configuration file at some point for stuff like draw speed, ip:port combinations, etc.

Comment (sneakers-the-rat) [reply vent/common/values.py:321]: but agreed with this until then^^

**Pull request merged**